### ISSUE TITLE
clarification on interaction between vkReleaseSwapchainImagesEXT and semaphore passed to vkAcquireNextImageKHR

### ISSUE THREAD

**Issue by AlexRouSg** [KhronosGroup/Vulkan-Docs#2278]
When using `vkReleaseSwapchainImagesEXT` to release the swapchain image, validation seems to think that the semaphore passed to `vkAcquireNextImageKHR` is signalled and I get the following validation error:
> Semaphore must not be currently signaled. The Vulkan spec states: If semaphore is not VK_NULL_HANDLE it must be unsignaled (https://vulkan.lunarg.com/doc/view/1.3.268.0/windows/1.3-extensions/vkspec.html#VUID-vkAcquireNextImageKHR-semaphore-01286)

I am unable to find any spec wording on what state should the semaphore be in after releasing said image and would like some clarification on what is the expected way to handle said semaphore after releasing the image.

**Comment by oddhack** [comment]
Tagged so SI TSG will pick this up.

**Comment by mizvekov** [comment]
Ping

**Comment by cubanismo** [comment]
@mizvekov, can you clarify your workflow a bit? It's unclear what may be causing the semaphore to appear in the improper state to validation, but I was wondering if you're doing something like the following:

VkSemaphore sem = createSemaphore();
vkAcquireNextImageKHR(..., &image_idx, sem);
...
vkReleaseSwapchainImagesEXT(..., &image_idx);
...
vkAcquireNextImageKHR(..., &image_idx, sem);

If that were the case, nothing would have reset sem between the first and second acquires (Releasing the images does not affect the state of the semaphore that came with it from vkAcquireNextImageKHR), so the validation error would be valid. However, it's hard to tell what could be going wrong without understanding more about how the semaphore is used prior to the problematic vkAcquireNextImageKHR call.

**Comment by spencer-lunarg** [comment]
There is a chance Validation Layers are just not tracking `VK_EXT_swapchain_maintenance1` correctly... I see we have no "positive test" to use `vkReleaseSwapchainImagesEXT` correctly... if someone could provide how to use it (or honestly write it as a test in VVL) it might show that this is a Validation Layer issue

**Comment by spencer-lunarg** [comment]
correction, we have [this test](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/b24c118732afddbbcc5a98583841c59d8e041042/tests/unit/wsi_positive.cpp#L1657) not sure what we need to do to modify it to reproduce your case

**Comment by AlexRouSg** [comment]
@cubanismo so the expected flow here is that if vkAcquireNextImageKHR returns >= 0, then even if vkReleaseSwapchainImagesEXT was called, the semaphore can only be waited on in a vkQueuSubmit or destroyed?

**Comment by mizvekov** [comment]
@cubanismo So suppose you need to rebuild the swap chain immediately after calling vkAcquireNextImage, for example because the window was resized, you got SUBOPTIMAL, etc.

It seems the only ways out of that are to wait on the semaphore, or to recycle the semaphore, which are cumbersome operations.

It would definitely be an improvement to usability if vkReleaseSwapchainImages would leave the semaphore in a state where you could use it again on vkAcquireNextImage.

I hope that clarifies it, but please let me know if you need more details.

**Comment by cubanismo** [comment]
>  the expected flow here is that if vkAcquireNextImageKHR returns >= 0, then even if vkReleaseSwapchainImagesEXT was called, the semaphore can only be waited on in a vkQueuSubmit or destroyed?

Correct. The semaphore is not directly associated with the image, so releasing it doesn't affect the semaphore. The semaphore just refers to work that needs to be completed before using the image again. Even if you don't use the image, that semaphore state still needs to be kept consistent somehow.

Note destroying a semaphore with an in-flight signal or wait is also invalid usage. Effectively, the semaphore must be waited for at some point, even if you just want to destroy it. I believe it would also be sufficient to wait on a fence returned from the same AcquireNextImageKHR call as the semaphore, since they should effectively reference the same work, but I'm not sure the spec or validation layers account for that.

> It seems the only ways out of that are to wait on the semaphore, or to recycle the semaphore, which are cumbersome operations.

Agreed, but vkReleaseSwapchainImages[EXT,KHR] is not a queue operation and doesn't accept any semaphores as parameters, so it has no logical way to alter any semaphore state.

**Comment by mizvekov** [comment]
> I believe it would also be sufficient to wait on a fence returned from the same AcquireNextImageKHR call as the semaphore, since they should effectively reference the same work, but I'm not sure the spec or validation layers account for that.

Yes, I assumed (but didn't state) that waiting on that fence would be necessary.

If the spec and validation layers are changed to agree on that, that's a perfectly good solution, and seems to match what implementations already do.

**Comment by cubanismo** [comment]
OK. I think updating this text in the "Implicit Synchronization Guarantees" section of the spec:

> The fence signal operation defined by the fence parameter of a [vkQueueSubmit](https://docs.vulkan.org/spec/latest/chapters/cmdbuffers.html#vkQueueSubmit) or [vkQueueSubmit2](https://docs.vulkan.org/spec/latest/chapters/cmdbuffers.html#vkQueueSubmit2) or [vkQueueBindSparse](https://docs.vulkan.org/spec/latest/chapters/sparsemem.html#vkQueueBindSparse) command is ordered after all semaphore signal operations defined by that command.

To include vkAcquireNextImageKHR and vkAcquireNextImage2KHR would be sufficient then. I'll propose that spec change.

**Comment by oddhack** [comment]
This should be fixed in the 1.4.343 spec update.
